docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,650 @@
+with remote metadata even if there is some local metadata that could be brought up to
+date by ingesting the WAL, because if another node updated
+the remote storage, it might have advanced the remote_consistent_lsn such that the safekeeper drops WAL
+content that would be needed for this node to recover. Reconcilation with remote metadata also acts as an optimization, to avoid redundantly
+re-ingesting WAL data that is already in S3.
+
+Because index files are now suffixed with generation numbers, some changes are needed
+to load them:
+
+- In a clean migration of a tenant, the control plane may remember the tenant's most recent
+  generation numbers, and provide them to the new node in the attach request. This is sufficient
+  for the new node to directly GET the latest index.
+- Alternatively, we may use the storage broker for this: if we are continuously publishing the generation
+  information to the broker, then a newly attached tenant may query this information back and if
+  the attachment generation is exactly 1 less than the current attachment, or if the attachment is
+  the same but the node generation is exactly 1 less and the node ID is the same, then we may use that
+  information to calculate the most recent index key. See "Optional Safekeeper changes" below for where
+  we publish generation info to the safekeeper & in turn to the storage broker.
+- As a fallback, newly attached tenants may issue a ListObjectsv2 request using index_part as a prefix
+  to enumerate the indices and pick the most recent one by attachment generation. The listing would
+  typically only return 1-2 indices.
+
+Once the new attachment has written out its index_part.json, it may clean up historic index_part.json
+files that were found, unless the control plane has indicated to us that the tenant is multiply attached
+(see the subsequent HA RFC for this concept).
+
+#### Node Startup with existing Attachment
+
+During attachment, we must also write out the attachment generation number to local storage,
+so that on subsequent node startup we can tell if our attachment is still current. On node startup, when we see an already-attached tenant, it is possible that the tenant was attached to a different node while this node was offline.
+We may still safely start ingesting data and serving reads, but we may not process deletions: deletions
+will be blocked when the deletion queue checks the attachment's generation and finds that it is
+not the latest. We may continue operating the attachment as normal until the control plane
+asks us to detach: while we are not the latest generation, it is possible that some endpoint
+is still configured to use us for reads, so we should continue operating until the control plane
+asks us to detach.
+
+On a new attachment (i.e. we do not have local disk state that was written within the attachment's
+generation number), we must recover metadata from remote storage. We can find the latest index_part.json as follows:
```

Comment:
> as follows: 

Seems like you re-ordered paragraphs and this one was further above before that?

Reply:
Cleaned this up in 590e50cd12fbee17dd1d9b13e552e206bd6faa8e